src/Aspire.Hosting/Dashboard/DashboardViewModelService.cs

```diff
@@ -294,6 +321,90 @@ private async Task WriteChange(ResourceViewModel resourceViewModel, ObjectChange
+                // the endpoint is ready
+                var content = await response.Content.ReadAsStringAsync(_cancellationToken).ConfigureAwait(false);
+                if (!string.Equals(content, errorContent, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                // Endpoint returned a valid response
+                return true;
+            }
+        }
+        catch
+        {
+            // Catch exceptions, we'll retry again
+        }
+
+        return false;
+    }
+
+    // Poll a given service endpoint until it doesn't return a bad 404 response from the proxy (or the attempt times out)
+    private async Task PollUntilServiceEndpointValid(WatchEventType eventType, Service service, TimeSpan maxRetryDuration)
+    {
+        try
+        {
+            const string errorContent = "404 page not found\n";
+
+            if (service.UsesHttpProtocol(out var uriScheme)
+                && !_checkedServices.ContainsKey(service.Metadata.Name))
+            {
+                // We need to verify if service is "actually" working
+                if (!string.Equals(service.Status?.State, "Ready", StringComparison.OrdinalIgnoreCase))
+                {
+                    // If service is not ready, we don't try to connect. We will get future notification that service moved to ready state
+                    return;
+                }
+
+                var currentTimestamp = DateTime.UtcNow;
+                var delay = TimeSpan.FromMilliseconds(40);
+                var proxyUrlString = $"{uriScheme}://{service.AllocatedAddress}:{service.AllocatedPort}";
+                while (true)
+                {
+                    if (await CheckServiceEndpointReady(proxyUrlString, errorContent).ConfigureAwait(false))
+                    {
+                        break;
+                    }
+
+                    if (DateTime.UtcNow.Subtract(currentTimestamp) > maxRetryDuration)
+                    {
+                        // We went over max retry duration so exit while
+                        _logger.LogDebug("Couldn't confirm {ServiceName} endpoint ready in {TimeoutSeconds}, assuming ready", service.Metadata.Name, maxRetryDuration.TotalSeconds);
+                        break;
+                    }
+
+                    await Task.Delay(delay, _cancellationToken).ConfigureAwait(false);
+                    delay *= 2;
+                }
+
+                _checkedServices.TryAdd(service.Metadata.Name, null);
+            }
+
+            await _kubernetesChangesChannel.Writer.WriteAsync(
+                (eventType, service.Metadata.Name, service), _cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogDebug("Failed to confirm {ServiceName} endpoint ready: {Error}", service.Metadata.Name, ex.Message);
```

Comment:
```suggestion
            _logger.LogDebug(ex, "Failed to confirm {ServiceName} endpoint ready", service.Metadata.Name);
```

Reply:
@danegsta FYI log level and logging an exception are independent of each other. We still want to log the full exception details. The change is to lower the log level (`LogError` -> `LogDebug`) so it isn't shown by default.

Reply:
Good to know! We've got similar levels implementation to the log library we use in the orchestrator, but it treats error as a special case for taking an exception.